dotcom-rendering/src/server/dev-index.html

```diff
@@ -657,6 +807,267 @@ <h2>${title}</h2>
 						}
 					}
 				});
+
+			class ArticleLink extends HTMLElement {
+				/**
+				 * @typedef {'dotcom' | 'apps' | 'amp'} Product
+				 * @typedef {'PROD' | 'CODE' | 'DEV'} Env
+				 */
+
+				/** @type {DocumentFragment} */
+				content;
+
+				/** @type {HTMLAnchorElement} */
+				link;
+
+				/** @type {HTMLSpanElement} */
+				linkText;
+
+				/**
+				 * @param {Product} product
+				 * @return {string}
+				 */
+				prodOrigin(product) {
+					switch (product) {
+						case 'dotcom':
+						case 'apps':
+							return 'https://www.theguardian.com';
+						case 'amp':
+							return 'https://amp.theguardian.com';
+					}
+				}
+
+				/**
+				 * @param {Product} product
+				 * @return {string}
+				 */
+				codeOrigin(product) {
+					switch (product) {
+						case 'dotcom':
+						case 'apps':
+							return 'https://m.code.dev-theguardian.com';
+						case 'amp':
+							return 'https://amp.code.dev-theguardian.com';
+					}
+				}
+
+				/**
+				 * @param {Product} product
+				 * @return {string}
+				 */
+				params(product) {
+					return product === 'apps' ? '?dcr=apps' : '';
+				}
+
+				/**
+				 * @param {Product} product
+				 * @return {string}
+				 */
+				devPath(product) {
+					switch (product) {
+						case 'dotcom':
+							return '/Article';
+						case 'apps':
+							return '/AppsArticle';
+						case 'amp':
+							return '/AMPArticle';
+					}
+				}
+
+				/**
+				 * @param {Product} product
+				 * @param {Env} env
+				 * @param {URL} url
+				 */
+				buildUrl(product, env, url) {
+					switch (env) {
+						case 'PROD':
+							return `${this.prodOrigin(product)}${
+								url.pathname
+							}${this.params(product)}`;
```

Comment:
We should use the `URL` constructer here for our return type to be correct:

```suggestion
							return new URL(`${
								url.pathname
							}${this.params(product)}`, this.prodOrigin(product)).href;
```

Reply:
Yeah, in fact several of these methods could be refactored to use the `URL` class. I didn't do this because I don't think we'd be making much use of the API; mostly just constructing with `string`s and then pulling similar `string`s back out.

However, I agree the types are more accurate so can update 👍.